File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/java/io/github/lishangbu/avalon/dataset/service/impl/MoveDamageClassServiceImplTest.java
package io.github.lishangbu.avalon.dataset.service.impl;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.github.lishangbu.avalon.dataset.entity.MoveDamageClass;
import io.github.lishangbu.avalon.dataset.repository.MoveDamageClassRepository;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

@ExtendWith(MockitoExtension.class)
class MoveDamageClassServiceImplTest {
    @Mock private MoveDamageClassRepository moveDamageClassRepository;
    @InjectMocks private MoveDamageClassServiceImpl moveDamageClassService;

    @Test
    void getPageByCondition_callsRepository() {
        MoveDamageClass moveDamageClass = new MoveDamageClass();
        Pageable pageable = PageRequest.of(0, 5);
        Page<MoveDamageClass> page = new PageImpl<>(List.of(moveDamageClass));
        when(moveDamageClassRepository.findAll(any(Example.class), eq(pageable))).thenReturn(page);

        Page<MoveDamageClass> result =
                moveDamageClassService.getPageByCondition(moveDamageClass, pageable);

        assertSame(page, result);
        ArgumentCaptor<Example> captor = ArgumentCaptor.forClass(Example.class);
        verify(moveDamageClassRepository).findAll(captor.capture(), eq(pageable));
        assertNotNull(captor.getValue().getProbe());
    }

    @Test
    void listByCondition_callsRepository() {
        MoveDamageClass moveDamageClass = new MoveDamageClass();
        List<MoveDamageClass> expected = List.of(moveDamageClass);
        when(moveDamageClassRepository.findAll(any(Example.class))).thenReturn(expected);

        List<MoveDamageClass> result = moveDamageClassService.listByCondition(moveDamageClass);

        assertSame(expected, result);
        verify(moveDamageClassRepository).findAll(any(Example.class));
    }

    @Test
    void save_usesRepository() {
        MoveDamageClass moveDamageClass = new MoveDamageClass();
        when(moveDamageClassRepository.save(moveDamageClass)).thenReturn(moveDamageClass);

        MoveDamageClass result = moveDamageClassService.save(moveDamageClass);

        assertSame(moveDamageClass, result);
        verify(moveDamageClassRepository).save(moveDamageClass);
    }

    @Test
    void update_usesRepository() {
        MoveDamageClass moveDamageClass = new MoveDamageClass();
        when(moveDamageClassRepository.save(moveDamageClass)).thenReturn(moveDamageClass);

        MoveDamageClass result = moveDamageClassService.update(moveDamageClass);

        assertSame(moveDamageClass, result);
        verify(moveDamageClassRepository).save(moveDamageClass);
    }

    @Test
    void removeById_callsRepository() {
        moveDamageClassService.removeById(1L);
        verify(moveDamageClassRepository).deleteById(1L);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/kotlin/io/github/lishangbu/avalon/dataset/service/impl/MoveDamageClassServiceImplTest.kt
package io.github.lishangbu.avalon.dataset.service.impl

import io.github.lishangbu.avalon.dataset.entity.MoveDamageClass
import io.github.lishangbu.avalon.dataset.entity.dto.MoveDamageClassSpecification
import io.github.lishangbu.avalon.dataset.entity.dto.SaveMoveDamageClassInput
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateMoveDamageClassInput
import io.github.lishangbu.avalon.dataset.repository.MoveDamageClassRepository
import org.babyfish.jimmer.Page
import org.babyfish.jimmer.sql.ast.mutation.AssociatedSaveMode
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import org.springframework.data.domain.PageRequest

class MoveDamageClassServiceImplTest {
    private val repository = mock(MoveDamageClassRepository::class.java)
    private val service = MoveDamageClassServiceImpl(repository)

    @Test
    fun getPageByCondition_callsRepository() {
        val specification = MoveDamageClassSpecification(id = "1", internalName = "physical")
        val pageable = PageRequest.of(0, 5)
        `when`(repository.findAll(specification, pageable)).thenReturn(
            Page(listOf(moveDamageClassEntity(1L, "physical", "物理", "desc")), 1, 1),
        )

        val result = service.getPageByCondition(specification, pageable)

        assertEquals(1, result.rows.size)
        assertEquals("1", result.rows.first().id)
    }

    @Test
    fun listByCondition_callsRepository() {
        val specification = MoveDamageClassSpecification(id = "1", internalName = "physical")
        `when`(repository.findAll(specification)).thenReturn(listOf(moveDamageClassEntity(1L, "physical", "物理", "desc")))

        val result = service.listByCondition(specification)

        assertEquals(1, result.size)
        assertEquals("1", result.first().id)
    }

    @Test
    fun save_usesInsertOnlyMode() {
        `when`(repository.save(any<MoveDamageClass>(), eq(SaveMode.INSERT_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())).thenReturn(moveDamageClassEntity(1L, "physical", "物理", "desc"))

        val result = service.save(SaveMoveDamageClassInput("physical", "物理", "desc"))

        assertEquals("1", result.id)
        verify(repository).save(any<MoveDamageClass>(), eq(SaveMode.INSERT_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())
    }

    @Test
    fun update_usesUpsertMode() {
        `when`(repository.save(any<MoveDamageClass>(), eq(SaveMode.UPSERT), eq(AssociatedSaveMode.REPLACE), isNull())).thenReturn(moveDamageClassEntity(1L, "special", "特殊", "desc"))

        val result = service.update(UpdateMoveDamageClassInput("1", "special", "特殊", "desc"))

        assertEquals("1", result.id)
        verify(repository).save(any<MoveDamageClass>(), eq(SaveMode.UPSERT), eq(AssociatedSaveMode.REPLACE), isNull())
    }

    @Test
    fun removeById_callsRepository() {
        service.removeById(1L)

        verify(repository).deleteById(1L)
    }
}

private fun moveDamageClassEntity(
    id: Long,
    internalName: String,
    name: String,
    description: String,
): MoveDamageClass =
    MoveDamageClass {
        this.id = id
        this.internalName = internalName
        this.name = name
        this.description = description
    }
